### PR TITLE
fix issue in toggleConsumption, #129

### PR DESCRIPTION
**Pull request by patriknw**
* the `require(!buffer.hasNext)` failed in poll
  if commit was requested when there was no demand and
  still things in the buffer from previous poll
* the reason was that we cached the assignments,
  they are automatically changed when using dynamic
  subscriptions and we must use the current value to
  resume/pause

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #129

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala [hunk 1]
```diff
@@ -360,15 +359,17 @@ private[kafka] abstract class ConsumerStageLogic[K, V, Out](
   protected def poll(): Unit = {
     try {
       def toggleConsumption() = {
+        // note that consumer.assignment() is automatically changed when using
+        // dynamic subscriptions and we must use the current value to resume/pause
         if (isAvailable(out))
-          consumer.resume(assignments: _*)
+          consumer.resume(consumer.assignment().asScala.toArray: _*)
```
Comment (drewhk) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: *ick*, that conversion is really ugly. Isn't this costly?

**Comment by drewhk** [comment]
LGTM

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala (hunk 1 above)
Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: yeah, that was why I optimized it previously (this was the original version), but that was before knowing how this actually worked.

Perhaps we can hand-roll the conversion to save the intermediate scala representation. In the end it's a varargs call. Stupid api.

**Comment by 13h3r** [comment]
No changes for `require(!buffer.hasNext)`?

**Comment by patriknw** [comment]
That `require` is still valid as far as I can see. Now we properly `pause` the topics when there is no demand and therefore the received messages from `poll` is 0.

**Comment by 13h3r** [comment]
May be we should take care about rebalance listener and update these topics?

**Comment by patriknw** [comment]
@13h3r what do you mean by "update these topics"? Are you thinking of that as a way around the varargs conversions?

**Comment by 13h3r** [comment]
@patriknw 

The `require` fails because we got some messages during `poll`. We got them because we did not pause topics. We did call `pause` and some topics are not paused. This means list of topics/partitions was changed after we gathered them. I may be wrong, so correct me if any. 

If all above is true we have two ways - always get and convert these topics or keep them actual in case of changing. I am talking about subscription to topics/partitions changes and updating this list of topics to pause/resume.

If I am wrong it will be a fail :)

**Comment by patriknw** [comment]
That is exactly the bug fix I have done here. The problem was that I grabbed the assignments once in the constructor. The fix is to grab them from the consumer every time, as in your original code.

We can work on performance improvement via `RebalanceListener` in a separate issue/pr.

**Pull request merged**